packages/core/src/index.tsx

```diff
@@ -0,0 +1,102 @@
+import { Context, Element } from 'koishi'
+import { } from 'koishi-plugin-puppeteer'
+
+// A helper function to mimic the behaviour of `styled-components`.
+// So we could use VSCode extension `vscode-styled-components` to
+// highlight our CSS code.
+const css = (args: TemplateStringsArray, ...values: any[]) => String.raw({ raw: args }, ...values)
+
+export function generateColor(color: WordleCore.BaseState): Partial<CSSStyleDeclaration> {
+  const base = {
+    color: '#fff',
+    border: '2px solid transparent',
+  }
+  switch (color) {
+    case 'correct':
+      return { ...base, backgroundColor: '#79b851' }
+    case 'wrong-place':
+      return { ...base, backgroundColor: '#f3c237' }
+    case 'none':
+      return { ...base, backgroundColor: '#a4aec4' }
+  }
+}
+
+const transformers: WordleCore.Transformers = {
+  'correct': (word) => `[${word}]`,
+  'wrong-place': (word) => `(${word})`,
+  'none': (word) => ` ${word} `,
+}
+
+export abstract class WordleCore<T extends WordleCore.Character = WordleCore.Character> {
+  protected todayWord: string
+
+  constructor(protected ctx: Context, protected config: WordleCore.Config) {
+    ctx.i18n.define('zh', require('./locales/zh-CN'))
+  }
+
+  abstract getTodayWord(): Promise<string>
+  abstract validateInput(input: string, solution: string): T[]
+  abstract getRandomWord(): string
+  public render(chars: T[][]): Element {
+    const textMode = !(this.config.imageMode && this.ctx.puppeteer)
```

Comment:
Better to print a warning message says something like `WARN: No puppeteer installed, switch to text mode.`